domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,209 @@
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.oppia.android.app.model.ScreenName
+import org.oppia.android.util.logging.ConsoleLogger
+import org.oppia.android.util.logging.ExceptionLogger
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.AppIconification
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.Snapshot
+import org.oppia.android.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+
+/**
+ * Snapshotter that gracefully and sequentially logs cpu usage across foreground and background
+ * moves of the application.
+ */
+class CpuPerformanceSnapshotter private constructor(
+  private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  private val performanceMetricsLogger: PerformanceMetricsLogger,
+  private val initialIconification: AppIconification,
+  private val consoleLogger: ConsoleLogger,
+  private val exceptionLogger: ExceptionLogger,
+  private val foregroundCpuLoggingTimePeriodMillis: Long,
+  private val backgroundCpuLoggingTimePeriodMillis: Long,
+  private val performanceMetricsAssessor: PerformanceMetricsAssessor
+) {
+
+  private val commandQueue by lazy { createCommandQueueActor() }
+
+  /** Updates the current [AppIconification] in accordance with the app's state changes. */
+  fun updateAppIconification(newIconification: AppIconification) {
+    sendSwitchIconificationCommand(newIconification)
+  }
+
+  private fun createCommandQueueActor(): SendChannel<CommandMessage> {
+    var currentIconification = initialIconification
+    var previousSnapshot =
+      performanceMetricsAssessor.computeSnapshotAtCurrentTime(currentIconification)
+    var switchIconificationCount = 0
+    val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+    return coroutineScope.actor<CommandMessage>(capacity = Channel.UNLIMITED) {
+      for (message in channel) {
+        when (message) {
+          is CommandMessage.SwitchIconification -> {
+            ++switchIconificationCount
+            sendLogSnapshotDiffCommand(
+              performanceMetricsAssessor.getRelativeCpuUsage(
+                previousSnapshot,
+                performanceMetricsAssessor.computeSnapshotAtCurrentTime(currentIconification)
+              )
+            )
+            currentIconification = message.newIconification
+            previousSnapshot =
+              performanceMetricsAssessor.computeSnapshotAtCurrentTime(currentIconification)
+            sendScheduleTakeSnapshotCommand(currentIconification, switchIconificationCount)
+          }
+          is CommandMessage.ScheduleTakeSnapshot -> scheduleTakeSnapshot(
+            message.currentIconification,
+            switchIconificationCount
+          )
+          is CommandMessage.TakeSnapshot -> {
+            if (message.switchId == switchIconificationCount) {
```

Comment:
Suggest making sure there's a test to verify the case when this if-check is necessary.

Reply:
Isn't this getting verified in the tests where we verify sequential logging?